core/include/detray/propagator/propagator.hpp

```diff
@@ -142,6 +142,41 @@ struct propagator {
         // Pass on the whether the propagation was successful
         return propagation._navigation.is_complete();
     }
+
+    template <typename state_t, typename actor_states_t = actor_chain<>::state>
+    DETRAY_HOST_DEVICE bool propagate_sync(state_t &propagation,
+                                           actor_states_t &&actor_states = {}) {
+
+        // Initialize the navigation
+        propagation._heartbeat = _navigator.init(propagation);
+
+        // Run all registered actors/aborters after init
+        run_actors(actor_states, propagation);
+
+        while (propagation._heartbeat) {
+
+            while (propagation._heartbeat) {
+
+                // Take the step
+                propagation._heartbeat &= _stepper.step(propagation);
+
+                // And check the status
+                propagation._heartbeat &= _navigator.update(propagation);
+
+                if (propagation._navigation.status() !=
```

Comment:
I must admit, I still don't fully understand this solution :). How does this actually reduce thread divergence? Tracks may very well already be on a sensitive surface after the first step they take and then, they move directly to l. 174 anyway, or is there some implicit barrier?

Reply:
>  Tracks may very well already be on a sensitive surface after the first step they take and then

Hmm Is this true? Couldn't some of them take more than one step due to overstepping?

Reply:
It is unpredictable.
My point was that: a track will get out of the internal while loop only when it is on the sensitive surface (let's assume that heartbeat is always true). and a thread in a warp won't execute L174 until all the other threads get out of the loop as well.
So the downside of this approach is clear because there will be a divergence when a certain thread is experiencing too many overstepping - I just want to make a fair comparison between `propagate` and `propagte_sync` in various cases.


Of course,  `propagate()` and `propagate_sync()` should not show any difference in CPU

Reply:
> but overstepping should be rare

Yeah it would be rare for "low" curvature tracks

Reply:
> is there some implicit barrier?

There is always a barrier for threads in a warp

Reply:
>  So the downside of this approach is clear because there will be a divergence when a certain thread is experiencing too many overstepping
If there is a step size constraint set for the stepper, the number of steps towards the next sensitive surface will be unpredictable, even if there is no overstepping at all. So, would you not see thread divergence regardless? If there is a passive surface in between for some tracks, it gets even more complicated...

> Of course, `propagate()` and `propagate_sync()` should not show any difference in CPU

Let's see, because the actor chain is run twice for sensitive surfaces, right?

Reply:
If it is running twice, then it is my mistake. I am still debugging the code

Reply:
Ah, no you may be right!

Reply:
> If there is a step size constraint set for the stepper, the number of steps towards the next sensitive surface will be unpredictable, even if there is no overstepping at all. So, would you not see thread divergence regardless? If there is a passive surface in between for some tracks, it gets even more complicated...

Woh. hold your horses. I am not saying that propagate_sync() must be better than propagate(), and I need to see the performance for various cases.  Of course, We will definitely need to avoid using step size constraint for optimal performance.

Reply:
Ok, nvm, then :D I am just trying to understand what is happening